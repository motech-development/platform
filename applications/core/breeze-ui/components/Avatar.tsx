import { FC } from 'react';

const Avatar: FC<{
  alt: string;
  src: string;
  width?: number | 'auto';
}> = ({ alt, src, width = 'auto' }) => (
  <img className={`h-${width} w-${width} rounded-full`} src={src} alt={alt} />
);

export default Avatar;
