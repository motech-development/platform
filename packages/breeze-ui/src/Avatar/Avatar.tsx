import React, { FC, memo } from 'react';
import styled from 'styled-components';

interface IImageProps {
  $width: number;
}

const Image = styled.img<IImageProps>`
  ${({ $width }) => `
    display: block;
    width: ${$width ? `${$width}px` : 'auto'};
  `}
`;

interface IWrapperProps {
  $width: number;
}

const Wrapper = styled.div<IWrapperProps>`
  ${({ $width }) => `
    border-radius: 50%;
    display: inline-block;
    overflow: hidden;
    width: ${$width ? `${$width}px` : 'auto'};
  `}
`;

export interface IAvatarProps {
  alt: string;
  src: string;
  width?: number;
}

const Avatar: FC<IAvatarProps> = ({ alt, src, width = 0 }) => (
  <Wrapper $width={width}>
    <Image alt={alt} src={src} $width={width} />
  </Wrapper>
);

export default memo(Avatar);
