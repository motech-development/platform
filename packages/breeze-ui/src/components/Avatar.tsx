import { FC, forwardRef, ImgHTMLAttributes } from 'react';
import { classNames } from '../utils/className';

export interface IAvatarProps extends ImgHTMLAttributes<HTMLImageElement> {
  alt: string;
  src: string;
}

const Avatar: FC<IAvatarProps> = forwardRef<HTMLImageElement, IAvatarProps>(
  ({ alt, className = 'h-8 w-8', src, ...rest }, ref) => (
    <img
      className={classNames('rounded-full', className)}
      src={src}
      alt={alt}
      ref={ref}
      {...rest}
    />
  ),
);

export default Avatar;
