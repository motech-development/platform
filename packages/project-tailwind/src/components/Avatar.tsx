import { ComponentPropsWithRef } from 'react';

/** Avatar component properties */
export interface IAvatarProps extends ComponentPropsWithRef<'img'> {
  /** Alternative information text */
  alt: string;

  /** Image path */
  src: string;
}

/**
 * Display user avatar photo
 *
 * @param props - Component props
 *
 * @returns Avatar component
 */
export function Avatar(props: IAvatarProps) {
  const { alt, className = 'h-8 w-8', src, ...rest } = props;

  return <img {...{ alt, className, src }} {...rest} />;
}
