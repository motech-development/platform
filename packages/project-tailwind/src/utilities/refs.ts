import { LegacyRef, MutableRefObject, useMemo, useRef } from 'react';

/**
 * Check if ref is a mutable ref object
 *
 * @param ref - Reference object or function
 *
 * @returns If ref is a mutable ref object
 */
function isMutableRefObject<T>(
  ref: MutableRefObject<T> | LegacyRef<T>,
): ref is MutableRefObject<T | null> {
  return ref != null;
}

/** useMergeRefs arguments */
type TUseMergeRefsInput<T> = Array<MutableRefObject<T> | LegacyRef<T>>;

/**
 * Combines multiple refs into a single memoized callback instance
 *
 * @param refs - Reference object or function
 *
 * @returns Combined ref object
 */
export default function useMergeRefs<T>(...refs: TUseMergeRefsInput<T>) {
  const targetRef = useRef<T>(null);

  return useMemo(() => {
    if (refs.every((ref) => ref == null)) {
      return null;
    }

    refs.forEach((ref) => {
      if (typeof ref === 'function') {
        ref(targetRef.current);
      } else if (isMutableRefObject(ref)) {
        // eslint-disable-next-line no-param-reassign
        ref.current = targetRef.current;
      }
    });

    return targetRef;
  }, [refs]);
}
