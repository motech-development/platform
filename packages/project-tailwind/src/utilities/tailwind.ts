import clsx, { ClassValue } from 'clsx';

/** Alignment options */
export enum Alignment {
  /** Centre */
  CENTRE = 'centre',
  /** Left */
  LEFT = 'left',
  /** Right */
  RIGHT = 'right',
}

/** Alignment type */
export type TAlignment = `${Alignment}`;

/** Sizing options */
export enum Sizing {
  /** Large */
  LG = 'lg',
  /** Medium */
  MD = 'md',
  /** None */
  NONE = 'none',
  /** Small */
  SM = 'sm',
}

/** Sizing type */
export type TSizing = `${Sizing}`;

/** Theming options */
export enum Themes {
  /** Danger */
  DANGER = 'danger',
  /** Primary */
  PRIMARY = 'primary',
  /** Secondary */
  SECONDARY = 'secondary',
  /** Success */
  SUCCESS = 'success',
  /** Warning */
  WARNING = 'warning',
}

/** Theme type */
export type TTheme = `${Themes}`;

/** Class names options */
type TCreateStylesClassNames = {
  /** Universal classes to apply */
  classNames?: ClassValue[];
  /** Sizing specific classes */
  sizing?: {
    [name in Sizing]?: ClassValue[];
  };
  /** Theme specific classes */
  theme?: {
    [name in TTheme]?: ClassValue[];
  };
};

/**
 * Tailwind hook
 *
 * @param theme - The theme to apply
 * @param sizing - Sizing to apply
 *
 * @returns Tailwind helper utilities
 */
export function useTailwind(theme: TTheme, sizing: TSizing) {
  /**
   * Creates css styles to apply
   *
   * @param classNames - Classes to be applied
   *
   * @returns Class names
   */
  function createStyles(classNames: TCreateStylesClassNames) {
    let finalStyles: ClassValue[] = [];

    if (classNames.classNames) {
      finalStyles = [...classNames.classNames];
    }

    if (sizing && classNames.sizing) {
      const styles = classNames.sizing[sizing];

      if (!styles) {
        throw new Error('No classes defined for this sizing.');
      }

      finalStyles = [...finalStyles, ...styles];
    }

    if (theme && classNames.theme) {
      const styles = classNames.theme[theme];

      if (!styles) {
        throw new Error('No classes defined for this theme.');
      }

      finalStyles = [...finalStyles, ...styles];
    }

    return clsx(finalStyles);
  }

  /**
   * Creates text alignment classes
   *
   * @param alignment - Text alignment
   *
   * @returns Class names
   */
  function createTextAlignmentStyles(alignment: TAlignment) {
    switch (alignment) {
      case Alignment.CENTRE:
        return 'text-center';
      case Alignment.RIGHT:
        return 'text-right';
      case Alignment.LEFT:
      default:
        return 'text-left';
    }
  }

  return {
    createStyles,
    createTextAlignmentStyles,
  };
}
