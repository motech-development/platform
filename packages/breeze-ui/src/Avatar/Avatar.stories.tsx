import { withA11y } from '@storybook/addon-a11y';
import { number, text, withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import BaseStyles from '../BaseStyles/BaseStyles';
import Avatar from './Avatar';

const stories = storiesOf('Avatar', module);

stories.addDecorator(withA11y);
stories.addDecorator(withKnobs);

stories.add('Basic avatar', () => (
  <>
    <BaseStyles />
    <Avatar
      alt={text('Alt text', 'Avatar')}
      src={text(
        'Image Url',
        'https://www.gravatar.com/avatar/8801091e665fdac669daa63d32167b7b',
      )}
      width={number('Width', 0)}
    />
  </>
));
