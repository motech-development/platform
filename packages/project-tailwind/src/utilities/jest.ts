/* eslint-disable import/no-extraneous-dependencies */
import { render, RenderResult } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { UserEvent } from '@testing-library/user-event/dist/types/setup/setup';
import { ReactElement } from 'react';
import { Sizing, Themes } from './tailwind';

export const themes = [
  {
    theme: Themes.DANGER,
  },
  {
    theme: Themes.PRIMARY,
  },
  {
    theme: Themes.SECONDARY,
  },
  {
    theme: Themes.SUCCESS,
  },
  {
    theme: Themes.WARNING,
  },
];

export const sizing = [
  {
    size: Sizing.LG,
  },
  {
    size: Sizing.MD,
  },
  {
    size: Sizing.NONE,
  },
  {
    size: Sizing.SM,
  },
];

type TSetup = RenderResult & {
  user: UserEvent;
};

/**
 * Combines Testing Library render function with user events setup
 *
 * @param ui - React element to test
 *
 * @returns Rendered component and test user
 */
export function setup(ui: ReactElement): TSetup {
  return {
    user: userEvent.setup({
      delay: null,
    }),
    ...render(ui),
  };
}
