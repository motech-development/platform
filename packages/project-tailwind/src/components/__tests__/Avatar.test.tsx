import { setup } from '../../utilities/jest';
import { Avatar } from '../Avatar';

describe('Avatar', () => {
  it('should correctly render component when no styles are set', () => {
    const { asFragment } = setup(
      <Avatar
        alt="Alt text"
        src="https://www.gravatar.com/avatar/8801091e665fdac669daa63d32167b7b"
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });

  it('should correctly render component when custom styles are set', () => {
    const { asFragment } = setup(
      <Avatar
        className="h-auto w-auto"
        alt="Alt text"
        src="https://www.gravatar.com/avatar/8801091e665fdac669daa63d32167b7b"
      />,
    );

    expect(asFragment()).toMatchSnapshot();
  });
});
