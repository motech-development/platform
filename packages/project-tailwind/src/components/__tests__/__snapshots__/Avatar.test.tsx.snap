// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar should correctly render component when custom styles are set 1`] = `
<DocumentFragment>
  <img
    alt="Alt text"
    class="h-auto w-auto"
    src="https://www.gravatar.com/avatar/8801091e665fdac669daa63d32167b7b"
  />
</DocumentFragment>
`;

exports[`Avatar should correctly render component when no styles are set 1`] = `
<DocumentFragment>
  <img
    alt="Alt text"
    class="h-8 w-8"
    src="https://www.gravatar.com/avatar/8801091e665fdac669daa63d32167b7b"
  />
</DocumentFragment>
`;
